File: backend/app/core/upgrades/variants/explosive_bullets.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class ExplosiveBullets(UpgradeImplementation):
    id = 'explosive_bullets'
    name = 'Explosive Bullets'
    description = 'Bullets explode on impact dealing collision and area damage.'
    rarity = 'epic'
    type = 'variant'
    cost = 20

    def can_apply(self, state: dict) -> bool:
        applied_upgrades = state.get('applied_upgrades', [])
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        appliers.variants.set_active('bullet', 'ExplosiveBullet')

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/variants/explosive_bullets.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, RarityID, UpgradeTypeID } from '../../data/ID'

export const ExplosiveBulletsDef: UpgradeDef = {
  id: "explosive_bullets",
  name: "Explosive Bullets",
  description: "Bullets explode on impact dealing collision and area damage.",
  rarity: RarityID.Epic,
  upgradeType: UpgradeTypeID.Variant,
  cost: 20,
  targetClass: UpgradeTargetID.Bullet,
  variantClass: "ExplosiveBullet",
  replaces: ["homing_bullets", "buckshot_bullets"],
  specificAttackType: "bullet",
  stackable: false,
  incompatibleWith: ["homing_bullets", "buckshot_bullets"],
}

export class ExplosiveBullets extends Upgrade {}
